activerecord/lib/active_record/associations/has_many_through_association.rb

```diff
@@ -38,10 +38,12 @@ def concat_records(records)
       def insert_record(record, validate = true, raise = false)
         ensure_not_nested
 
-        if raise
-          record.save!(:validate => validate)
-        else
-          return unless record.save(:validate => validate)
+        if !record.readonly?
```

Comment:
My point is that `if record.changed?; record.save; end` won't work if the change would have occurred in a callback.